packages/core-blockchain/lib/blockchain.js

```diff
@@ -263,80 +264,81 @@ module.exports = class Blockchain {
 
         state.lastBlock = block
 
-        callback()
+        return callback()
       } else if (block.data.height > this.getLastBlock().data.height + 1) {
         state.lastDownloadedBlock = state.lastBlock
-        callback()
+        return callback()
       } else if (block.data.height < this.getLastBlock().data.height || (block.data.height === this.getLastBlock().data.height && block.data.id === this.getLastBlock().data.id)) {
         state.lastDownloadedBlock = state.lastBlock
-        callback()
+        return callback()
       } else {
         state.lastDownloadedBlock = state.lastBlock
         logger.info(`Block ${block.data.height.toLocaleString()} disregarded because on a fork :knife_fork_plate:`)
-        callback()
+        return callback()
       }
     } else {
       logger.warn(`Block ${block.data.height.toLocaleString()} disregarded because verification failed :scroll:`)
-      callback()
+      return callback()
     }
   }
 
   /**
    * Process the given block.
+   * We should be sure this is fail safe (ie callback() is being called)
    * @param  {Block} block
    * @param  {Function} callback
    * @return {(Function|void)}
    */
   async processBlock (block, callback) {
     if (!block.verification.verified) {
       logger.warn(`Block ${block.data.height.toLocaleString()} disregarded because verification failed :scroll:`)
-
       return callback()
     }
 
-    const state = stateMachine.state
```

Comment:
Missing tests for the changes from here to `return this.dispatch('FORK')`.

Reply:
i have just put the same logic to the calling function instead to be sure this is threadsafe.
IE moving logic from `acceptChainedBlock` to `processBlock`